common/djangoapps/track/backends/oli.py

```diff
@@ -0,0 +1,121 @@
+"""OLI analytics service event tracker backend."""
+
+from __future__ import absolute_import
+
+import json
+import logging
+
+from requests_oauthlib import OAuth1Session
+
+from django.contrib.auth.models import User
+
+from student.models import anonymous_id_for_user
+from track.backends import BaseBackend
+
+
+LOG = logging.getLogger(__name__)
+
+
+class OLIAnalyticsBackend(BaseBackend):
+    """
+    Transmit events to the OLI analytics service
+    """
+    def __init__(self, **kwargs):
+        super(OLIAnalyticsBackend, self).__init__(**kwargs)
+
+        # the oli analytics service endpoint will be at
+        # "{}{}".format(oli_analytics_service_url, path)
+        self.url = kwargs.get('url', None)
+        self.path = kwargs.get('path', None)
+
+        # shared key for OLI analytics service
+        self.key = kwargs.get('key', None)
+
+        # shared secret for OLI analytics service
+        self.secret = kwargs.get('secret', None)
+
+        self.oauth = OAuth1Session(self.key, client_secret=self.secret)
+
+        # only courses with id in this set will have their data sent
+        self.course_id_set = set(kwargs.get('course_ids', []))
+
+    def send(self, event):
+        """
+        Forward the event to the OLI analytics server
+        Exact API here: https://docs.google.com/document/d/1ZB-qwP0bV7ko_xJdJNX1PYKvTyYd4I8CBltfac4dlfw/edit?pli=1#
+        OAuth 1 with nonce and body signing
+        """
+        if not (self.url and self.secret and self.key):
+            return
+
+        # Only currently passing problem_check events, which are CAPA only
+        if event.get('event_type') != 'problem_check':
+            return
+
+        if event.get('event_source') != 'server':
+            return
+
+        context = event.get('context')
+        if not context:
+            return
+
+        course_id = context.get('course_id')
+        if not course_id or course_id not in self.course_id_set:
+            return
+
+        user_id = context.get('user_id')
+        if not user_id:
+            return
+
+        event_data = event.get('event')
+        if not event_data:
+            return
+
+        problem_id = event_data.get('problem_id')
+        if not problem_id:
+            return
+
+        success = event_data.get('success')
+        if not success:
+            return
+
+        is_correct = success == 'correct'
+
+        # put the most expensive operation (DB access) at the end, to not do it needlessly
+        try:
+            user = User.objects.get(pk=user_id)
+        except User.DoesNotExist:
+            return
+
+        payload = {
+            'course_id': course_id,
+            'resource_id': problem_id,
+            'student_id': anonymous_id_for_user(user, None),
+            'result': is_correct
+        }
+
+        request_payload_string = json.dumps({'payload': json.dumps(payload)})
+        request_payload = {'request': request_payload_string}
+        endpoint = self.url + self.path
```

Comment:
This is brittle and will break if the `url` lacks a trailing `/`. This
is non-obvious, nor is it documented. It will also break if `path` is
`None`. This is why we should never manually manipulate file paths or
URLs.  There are far too many edge cases. This is also why Python
provides the `os.path` and `urlparse` modules.

This should be written like:

```python
from urlparse import urljoin
# ...
endpoint = urljoin(self.url, self.path)
```